llmfoundry/models/mpt/modeling_mpt.py

```diff
@@ -216,6 +224,45 @@ def gen_attention_mask_in_length(sequence_id: Union[None, torch.Tensor], S: int,
     return attention_mask_in_length
 
 
+def gen_flash_attn_padding_info(
+        bsz: int,
+        S: int,
+        past_key_len: int,
+        device: torch.device,
+        attention_mask_in_length: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None):
+    flash_attn_padding_info = {}
+    if attention_mask_in_length is None:
+        key_padding_mask = attention_mask
+        if key_padding_mask is None:
+            key_padding_mask = torch.ones((bsz, past_key_len + S),
+                                          dtype=torch.bool).to(device=device)
+        query_padding_mask = key_padding_mask[:, -S:]
+        unpadding_function = bert_padding.unpad_input
+    else:
+        key_padding_mask = attention_mask_in_length
+        query_padding_mask = attention_mask_in_length
+        unpadding_function = bert_padding.unpad_input_for_concatenated_sequences
+
+    _, indices_q, cu_seqlens_q, max_seqlen_q = unpadding_function(
+        torch.zeros(bsz, S, 1).to(device=device), query_padding_mask)
+    _, indices_k, cu_seqlens_k, max_seqlen_k = unpadding_function(
+        torch.zeros(bsz, past_key_len + S, 1).to(device=device),
+        key_padding_mask)
+    _, indices_v, _, _ = unpadding_function(
+        torch.zeros(bsz, past_key_len + S, 1).to(device=device),
```

Comment:
```suggestion
            key_padding_mask = torch.ones((bsz, past_key_len + S),
                                          dtype=torch.bool, device=device)
        query_padding_mask = key_padding_mask[:, -S:]
        unpadding_function = bert_padding.unpad_input
    else:
        key_padding_mask = attention_mask_in_length
        query_padding_mask = attention_mask_in_length
        unpadding_function = bert_padding.unpad_input_for_concatenated_sequences

    _, indices_q, cu_seqlens_q, max_seqlen_q = unpadding_function(
        torch.zeros(bsz, S, 1, device=device), query_padding_mask)
    _, indices_k, cu_seqlens_k, max_seqlen_k = unpadding_function(
        torch.zeros(bsz, past_key_len + S, 1, device=device),
        key_padding_mask)
    _, indices_v, _, _ = unpadding_function(
        torch.zeros(bsz, past_key_len + S, 1, device=device),
```

instantiate on GPU instead of on CPU then move to GPU